File: backend/src/services/userService.js
/**
 * User Service - Firebase Realtime Database
 */

const bcrypt = require('bcryptjs');
const { v4: uuidv4 } = require('uuid');
const { db, getAll, getById, create, update, remove, queryByChild } = require('../config/firebase');

const USERS_REF = 'users';

/**
 * Remove sensitive data from user object
 */
const sanitizeUser = (user) => {
  if (!user) return null;
  const { passwordHash, ...safeData } = user;
  return safeData;
};

/**
 * Create a new user
 */
const createUser = async ({ email, password, name, role = 'user' }) => {
  // Check if email already exists
  const existing = await queryByChild(USERS_REF, 'email', email);
  if (existing.length > 0) {
    throw new Error('Email already registered');
  }

  const id = uuidv4();
  const passwordHash = await bcrypt.hash(password, 10);
  const createdAt = new Date().toISOString();

  const userData = {
    email,
    name,
    role,
    passwordHash,
    createdAt,
  };

  await create(USERS_REF, id, userData);
  return sanitizeUser({ id, ...userData });
};

/**
 * Get user by email (includes passwordHash for authentication)
 */
const getUserByEmail = async (email) => {
  const users = await queryByChild(USERS_REF, 'email', email);
  if (users.length === 0) return null;
  return users[0];
};

/**
 * Get all users (sanitized)
 */
const getUsers = async () => {
  const users = await getAll(USERS_REF);
  // Sort by createdAt DESC
  users.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
  return users.map(sanitizeUser);
};

/**
 * Get user by ID (sanitized)
 */
const getUserById = async (userId) => {
  const user = await getById(USERS_REF, userId);
  return sanitizeUser(user);
};

/**
 * Update user role
 */
const updateUserRole = async (userId, role) => {
  await update(USERS_REF, userId, { role });
  return getUserById(userId);
};

/**
 * Update user profile (name only for now)
 */
const updateUserProfile = async (userId, updates) => {
  const { name } = updates;
  if (name) {
    await update(USERS_REF, userId, { name });
  }
  return getUserById(userId);
};

/**
 * Change user password
 */
const changePassword = async (userId, currentPassword, newPassword) => {
  // Get user with password hash
  const user = await getById(USERS_REF, userId);
  if (!user) {
    throw new Error('User not found');
  }

  // Verify current password
  const isValid = await bcrypt.compare(currentPassword, user.passwordHash);
  if (!isValid) {
    throw new Error('Current password is incorrect');
  }

  // Hash new password and update
  const newPasswordHash = await bcrypt.hash(newPassword, 10);
  await update(USERS_REF, userId, { passwordHash: newPasswordHash });

  return true;
};

/**
 * Delete user by ID
 */
const deleteUserById = async (userId) => {
  return remove(USERS_REF, userId);
};

module.exports = {
  createUser,
  getUserByEmail,
  getUsers,
  getUserById,
  updateUserRole,
  updateUserProfile,
  changePassword,
  deleteUserById,
};

File: frontend/lib/services/user_service.dart
import '../models/user.dart';
import 'api_client.dart';

class UserService {
  UserService._();

  static final UserService instance = UserService._();
  final ApiClient _client = ApiClient.instance;

  /// Update user profile (name)
  Future<UserModel> updateProfile({required String name}) async {
    final response = await _client.put('/users/profile', {'name': name});
    return UserModel.fromJson(response['user'] as Map<String, dynamic>);
  }

  /// Change password
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    await _client.post('/users/change-password', {
      'currentPassword': currentPassword,
      'newPassword': newPassword,
    });
  }

  /// Get all users (admin only)
  Future<List<UserModel>> fetchUsers() async {
    final response = await _client.get('/users');
    final users = response['users'] as List<dynamic>? ?? [];
    return users
        .map((json) => UserModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  /// Get user by ID (admin only)
  Future<UserModel> fetchUser(String userId) async {
    final response = await _client.get('/users/$userId');
    return UserModel.fromJson(response['user'] as Map<String, dynamic>);
  }

  /// Update user role (admin only)
  Future<UserModel> updateUserRole(String userId, String role) async {
    final response = await _client.patch('/users/$userId/role', {'role': role});
    return UserModel.fromJson(response['user'] as Map<String, dynamic>);
  }

  /// Delete user (admin only)
  Future<void> deleteUser(String userId) async {
    await _client.delete('/users/$userId');
  }
}
